doc/examples/contextual_enhancement.py

```diff
@@ -0,0 +1,275 @@
+PDE with evolution time :math:`t\geq 0` is given by:
+
+.. math::
+
+    \begin{cases}
+  \frac{\partial}{\partial t} W({\bf y},{\bf n},t) &= ((D^{33}({\bf n} \cdot
+          \nabla)^2 + D^{44} \Delta_{S^2})W)({\bf y},{\bf n},t)
+  \\ W({\bf y},{\bf n},0) &= U({\bf y},{\bf n})
+  \end{cases},
+
+where:
+
+* :math:`D^{33}>0` is  the coefficient for the spatial smoothing (which goes only in the direction of :math:`n`);
+
+* :math:`D^{44}>0` is the coefficient for the angular smoothing (here :math:`\Delta_{S^2}` denotes the Laplace-Beltrami operator on the sphere :math:`S^2`);
+
+* :math:`U:\mathbb{R}^3\rtimes S^2 \to \mathbb{R}` is the initial condition given by the noisy FOD/ODF’s field.
+
+This equation is solved via a shift-twist convolution (denoted by :math:`\ast_{\mathbb{R}^3\rtimes S^2}`) with its corresponding kernel :math:`P_t:\mathbb{R}^3\rtimes S^2 \to \mathbb{R}^+`:
+
+.. math::
+
+  W(\vec{y},\vec{n},t) = (P_t \ast_{\mathbb{R}^3 \rtimes S^2} U)(\vec{y},\vec{n})
+  = \int_{\mathbb{R}^3} \int_{S^2} P_t (R^T_{\vec{n}^\prime}(\vec{y}-\vec{y}^\prime),
+   R^T_{\vec{n}^\prime} \vec{n} ) U(\vec{y}^\prime, \vec{n}^\prime)
+
+Here, :math:`R_{\bf n}` is any 3D rotation that maps the vector :math:`(0,0,1)`
+onto :math:`{\bf n}`.
+
+Note that the shift-twist convolution differs from a Euclidean convolution and
+takes into account the non-flat structure of the space :math:`\mathbb{R}^3\rtimes S^2`.
+
+The kernel :math:`P_t` has a stochastic interpretation [DuitsAndFranken_JMIV]_.
+It can be seen as the limiting distribution obtained by accumulating random
+walks of particles in the position/orientation domain, where in each step the
+particles can (randomly) move forward/backward along their current orientation,
+and (randomly) change their orientation.  This is an extension to the 3D case of
+the process for contour enhancement of 2D images.
+
+.. figure:: _static/stochastic_process.jpg
+   :scale: 60 %
+   :align: center
+
+   The random motion of particles (a) and it's corresponding probability map (b) in 2D. The 3D kernel is shown on the right. Adapted from [Portegies2015_PLoSOne]_.
+
+In practice, as the exact analytical formulas for the kernel :math:`P_t`
+are unknown, we use the approximation given in [Portegies2015_SSVM]_.
+
+"""
+
+"""
+The enhancement is evaluated on the Stanford HARDI dataset
+(150 orientations, b=2000s/mm^2) where Rician noise is added. Constrained
+spherical deconvolution is used to model the fiber orientations.
+
+"""
+
+import numpy as np
+from dipy.data import fetch_stanford_hardi, read_stanford_hardi
+from dipy.sims.voxel import add_noise
+from dipy.core.gradients import gradient_table
+
+# Read data
+fetch_stanford_hardi()
+img, gtab = read_stanford_hardi()
+data = img.get_data()
+
+# Add Rician noise
+from dipy.segment.mask import median_otsu
+b0_mask, mask = median_otsu(data)
+np.random.seed(1)
+data_noisy = add_noise(data, 1, np.mean(data[mask]), noise_type='rician')
```

Comment:
It would make more sense to compute that only in the b0/wm, in the sense that the signal is higher than anywhere else and not influenced by the b-value. This way, your SNR will be higher and produce a larger variance for the noise, but still give a rough estimate between datasets, whereas if you want to compare a dataset with SNR computed at b1000 and b3000 it will be harder to relate the two together.